### ISSUE TITLE
SSR Compatibility (React/Gatsby)

### ISSUE THREAD

**Issue by mannp** [chiiya/haven#63]
Hi there

Firstly, thanks for creating :+1: 

Not sure if its a typo but _`import Haven from '@chiiya@haven'; `didn't work for me, but `import Haven from '@chiiya/haven';` did..

Resolved my reactjs issues, Monday mind block :)

**Comment by chiiya** [comment]
Thanks for the notice! I fixed the documentation.
Would you mind letting me know how you resolved your react issue? I would assume putting the haven instantiation in `componentDidMount` (or `useEffect`?) would be enough to make it work in an SSR context. I haven't used react myself, but it might be nice to include an excample in the docs.

**Comment by mannp** [comment]
Well I guess 'resolved' was a bit of a stretch :) as really I only managed to get the banner displaying by adding the Haven.create into my gatsby-browser.js and went to read more about gatsby-browser.js and gatsby-ssr.js ....

I tried #cookie-preferences and failed to get it to display anything though ..

**Comment by chiiya** [comment]
Okay. It could be an issue with the cookie preferences react component not being mounted yet when Haven tries to find it? I'm not sure about the best way to resolve this. I could expose a `getPreferences()` method on the Haven instance, then you could in theory do something like this in the cookie preferences component mounted hook:

```js
import { Haven } from '@chiiya/haven';

....

const instance = Haven.getInstance();
instance.getPreferences().init();
```

**Comment by mannp** [comment]
I am not familiar enough myself to suggest an optimal solution I'm afraid.

Willing to try anything you may have time to code, if that is any help.

**Comment by Grsmto** [comment]
SSR does not work because the package is using `window` and it's `undefined` on Node.
Common practice is to wrap every usage of the `window` object in a `if (isBrowser)` condition (which is something like `if (typeof window !== 'undefined')`.

**Comment by chiiya** [comment]
Sure, but Haven doesn't make any calls to `window` until you actually initialize it; there should be no side effects when you import it. Most frameworks offer some kind of lifecycle hooks that will only be executed client-side.
So for example in Vue any code inside `beforeMount` or `mounted` will not be executed in a SSR context. There shouldn't be any issues as long as the `Haven.create` call is inside one of these lifecycle hooks.

**Comment by Grsmto** [comment]
Actually it does have a side effect [here](https://github.com/chiiya/haven/blob/master/src/index.umd.ts#L4). At least that's the one Nextjs is complaining about which makes the lib not usable with it!

**Comment by chiiya** [comment]
Okay, I did not consider that for the server build the UMD bundle would be used. Makes sense though, since Node can't really deal too well with es-modules.
I've tried wrapping the offending code in the `typeof window` check you suggested, mind checking if it works now?
See [v0.6.1](https://github.com/chiiya/haven/releases/tag/v0.6.1)

**Comment by Grsmto** [comment]
Yes it works now! 👏

**Comment by Grsmto** [comment]
Ok so SSR do work fine. However I'm having the same issue of the `#cookie-preferences` not being server-side rendered and hence the plugin doesn't catch it.
I think an API like you proposed to have control on the initialisation of it would be great anyway!

**Comment by chiiya** [comment]
Should have been resolved with [v0.6.2](https://github.com/chiiya/haven/releases/tag/v0.6.2)
